### PR TITLE
Solarize_ops for Tensorflow Addons

### PR DESCRIPTION
**Pull request by vinayvr11**
See #1333

`solarize_ops` for image processing to add in Tensorflow Addons. Contains:
1. `solarize`
2. `solarize_add`

### CODE REVIEW

**Comment by abhichou4** [comment]
It would be nice to have tests for each op. Maybe add an image like #1338

**Comment by vinayvr11** [comment]
> It would be nice to have tests for each op. Maybe add an image like #1338

Sure

**Comment by vinayvr11** [comment]
@abhichou4 : Shall i need to test the images with all dtypes

**Comment by vinayvr11** [comment]
I am getting some errors @abhichou4 could you please help me out with this.

**Comment by abhichou4** [comment]
> I am getting some errors @abhichou4 could you please help me out with this.

Have a look at [contribution guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md) and add the pre-commit hooks. Also, BUILD and __init__.py need to be changed.

**Comment by vinayvr11** [comment]
> > I am getting some errors @abhichou4 could you please help me out with this.
> 
> Have a look at [contribution guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md) and add the pre-commit hooks. Also, [BUILD](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/BUILD) and [**init**.py](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/__init__.py) need to be changed.

@abhichou4 : Could you please review my commit.

File: tensorflow_addons/image/BUILD [hunk 1]
```diff
@@ -1,25 +1,28 @@
...
+            "distort_image_ops.py",
+            "filters.py",
+            "transform_ops.py",
+            "translate_ops.py",
+            "utils.py",
+            "sparse_image_warp.py",
+            "interpolate_spline.py",
+            "connected_components.py",
+            "resampler_ops.py",
+            "solarize_ops.py",
```
Comment (abhichou4) [inline tensorflow_addons/image/BUILD]: `compose_ops.py` is missing here.

**Comment by abhichou4** [comment]
> > > I am getting some errors @abhichou4 could you please help me out with this.
> > 
> > 
> > Have a look at [contribution guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md) and add the pre-commit hooks. Also, [BUILD](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/BUILD) and [**init**.py](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/__init__.py) need to be changed.
> 
> @abhichou4 : Could you please review my commit.

Running `tools/pre_commit.sh` will fix formatting fails. Please confirm if your tests work with `pytest`.

**Comment by vinayvr11** [comment]
> > > > I am getting some errors @abhichou4 could you please help me out with this.
> > > 
> > > 
> > > Have a look at [contribution guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md) and add the pre-commit hooks. Also, [BUILD](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/BUILD) and [**init**.py](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/__init__.py) need to be changed.
> > 
> > 
> > @abhichou4 : Could you please review my commit.
> 
> Running `tools/pre_commit.sh` will fix formatting fails. Please confirm if your tests work with `pytest`.
> Also, change your tests in accordance with #1328

Actually i am using windows so that command was not working it shows error. So you have any other options like pylint

**Comment by bhack** [comment]
> Actually i am using windows so that command was not working it shows error. So you have any other options like pylint

You are on Windows you need to run that in Docker.

I've prepared a devcontainer for Vscode ide+Vscode remote extension users but It Is still not merged. It could simplify the Dev environment also on Windows. You could give a try Is here:
 
https://github.com/tensorflow/addons/pull/1309

**Comment by vinayvr11** [comment]
> > Actually i am using windows so that command was not working it shows error. So you have any other options like pylint
> 
> You are on Windows you need to run that in Docker.
> 
> I've prepared a devcontainer for Vscode ide+Vscode remote extension users but It Is still not merged. It could simplify the Dev environment also on Windows. You could give a try Is here:
> 
> #1309

Thank you very much @bhack because for now i am using three commands separately black , pylint , flake8. So i need a way to do this fastly. Thank you.

File: tensorflow_addons/image/BUILD [hunk 2]
```diff
@@ -1,25 +1,28 @@
 licenses(["notice"])  # Apache 2.0
```
Comment (Squadrick) [inline tensorflow_addons/image/BUILD]: This is not how the `BUILD` is formatted. Please look at [Development tips](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#development-tips) to know how to auto-format your code. 

If you're on Windows, you can use docker to do the same.

File: tensorflow_addons/image/solarize_ops.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+""" This module is used to invert all pixel values above a threshold
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops.py]: Missing `LICENSE`, see any other source file to get the license.

File: tensorflow_addons/image/solarize_ops.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+   which simply means segmentation. """
+
+import tensorflow as tf
+
+
+def solarize(image, threshold=128):
+    """Method to solarize the image
+  image: input image
+  threshold: threshold value to solarize the image
+  """
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops.py]: Use proper indents.

File: tensorflow_addons/image/solarize_ops.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+""" This module is used to invert all pixel values above a threshold
+   which simply means segmentation. """
+
+import tensorflow as tf
+
+
+def solarize(image, threshold=128):
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops.py]: Missing type.

File: tensorflow_addons/image/solarize_ops.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    # Otherwise, subtract 255 from the pixel.
+    return tf.where(image < threshold, image, 255 - image)
+
+
+def solarize_add(image, addition=0, threshold=128):
+    """Method to add solarize to the image
+  image: input image
+  addition: addition amount to add in image
+  threshold: threshold value to solarize the image
+  """
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops.py]: Indent properly.

File: tensorflow_addons/image/solarize_ops.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  image: input image
+  threshold: threshold value to solarize the image
+  """
+    # For each pixel in the image, select the pixel
+    # if the value is less than the threshold.
+    # Otherwise, subtract 255 from the pixel.
+    return tf.where(image < threshold, image, 255 - image)
+
+
+def solarize_add(image, addition=0, threshold=128):
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops.py]: Missing type info.

File: tensorflow_addons/image/solarize_ops.py (hunk 4 above)
Comment (Squadrick) [reply tensorflow_addons/image/solarize_ops.py]: Look at other source files to see how docstrings are written.

File: tensorflow_addons/image/solarize_ops.py (hunk 6 above)
Comment (Squadrick) [reply tensorflow_addons/image/solarize_ops.py]: Look at other source files to see how docstrings are written.

File: tensorflow_addons/image/solarize_ops.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    """Method to add solarize to the image
+  image: input image
+  addition: addition amount to add in image
+  threshold: threshold value to solarize the image
+  """
+    # For each pixel in the image less than threshold
+    # we add 'addition' amount to it and then clip the
+    # pixel value to be between 0 and 255. The value
+    # of 'addition' is between -128 and 128.
+    added_image = tf.cast(image, tf.int64) + addition
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops.py]: Why the cast to `int64` instead of `int32`?

File: tensorflow_addons/image/solarize_ops.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+def solarize_add(image, addition=0, threshold=128):
+    """Method to add solarize to the image
+  image: input image
+  addition: addition amount to add in image
+  threshold: threshold value to solarize the image
+  """
+    # For each pixel in the image less than threshold
+    # we add 'addition' amount to it and then clip the
+    # pixel value to be between 0 and 255. The value
+    # of 'addition' is between -128 and 128.
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops.py]: Add a check to ensure `addition` is between -128 and 128.

File: tensorflow_addons/image/solarize_ops_test.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+"""Test of solarize_ops"""
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops_test.py]: Missing LICENSE.

File: tensorflow_addons/image/solarize_ops_test.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import sys
+import pytest
+import tensorflow as tf
+from tensorflow_addons.image import solarize_ops
+from tensorflow_addons.utils import test_utils
+from absl.testing import parameterized
+
+
+@test_utils.run_all_in_graph_and_eager_modes
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops_test.py]: See #1328

File: tensorflow_addons/image/solarize_ops_test.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+class SolarizeOPSTest(tf.test.TestCase, parameterized.TestCase):
+    """SolarizeOPSTest class to test the solarize images"""
+
+    def test_solarize(self):
```
Comment (Squadrick) [inline tensorflow_addons/image/solarize_ops_test.py]: Add a test for `solarize_add`

**Comment by seanpmorgan** [comment]
> Actually i am using windows so that command was not working it shows error. So you have any other options like pylint

@vinayvr11 There is a section for running pre-commit with Windows. Could you expand on the error you're getting and we'll be sure to update the steps accordingly:
https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style

**Comment by bhack** [comment]
This is now in https://github.com/tanzhenyu/image_augmentation /cc @tanzhenyu

**Pull request closed without merging**